# Clustering Neighbourhoods in Toronto.

# Part I. Web Scraping.

#### Installation of the required packages.

In [3]:
!pip install opencage
#!conda install beautifulsoup4
#!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    altair-4.0.1               |             py_0         575 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.0 MB

The following NEW packages will be 

#### Import of the required modules.

In [4]:
import lxml
import folium
import requests
import numpy as np
import pandas as pd
import matplotlib.cm as cm
import matplotlib.colors as colors

from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
from pandas.io.json import json_normalize
from opencage.geocoder import OpenCageGeocode

#### Getting the Neighbourhoods table from the Wikipedia page.

In [5]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
soup = BeautifulSoup(requests.get(url).content,features = 'lxml')
table = soup.find_all('table',class_ = 'wikitable sortable')[0]
df = pd.read_html(str(table))[0]

#### Transforming the data into the required form.

In [6]:
df = df.loc[(df['Borough'] != 'Not assigned')]
mask = (df['Neighbourhood'] == 'Not assigned')
df['Neighbourhood'][mask] = df['Borough'][mask]
df = df.sort_values(['Postcode','Borough','Neighbourhood'])
df = df.groupby(['Postcode','Borough'],as_index = False).agg({'Neighbourhood':', '.join})
df.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Malvern, Rouge
1      M1C  Scarborough  Highland Creek, Port Union, Rouge Hill
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

#### Printing the number of rows and columns of the resulting dataframe.

In [7]:
print("The number of rows and columns of the resulting dataframe:",df.shape)

The number of rows and columns of the resulting dataframe: (103, 3)


# Part II. Getting the coordinates for each neighbourhood.

#### Getting the geocodes for each postcode.

As neither "geopy.geocoders" nor "geocoder.google" did not work for me; I started to look for something else.
After some googling, I've found "opencage.geocoder".  
This worked well, excluding three cases, where opencage did not find the postcodes: M3A, M7A, M7R. 
I've updated their latitudes and longitudes manually.

In [8]:
# The code was removed by Watson Studio for sharing.

In [9]:
geocoder = OpenCageGeocode(key)
latitude_list = []
longitude_list = []
response = []

for i in df.index:
    address = df['Postcode'][i] + ', Toronto, Ontario'   
    results = geocoder.geocode(address) 
    for j in range(0,len(results)):
        if df['Postcode'][i] in results[j]['formatted']:
            latitude = results[j]['geometry']['lat']
            longitude = results[j]['geometry']['lng']
            break
        else:
            latitude = 0
            longitude = 0
    response.append(results[j]['formatted'])            
    latitude_list.append(latitude)
    longitude_list.append(longitude)
    #print('The geograpical coordinate are {}, {}.'.format(latitude, longitude))

df['Latitude'] = latitude_list
df['Longitude'] = longitude_list
df['response'] = response

df.loc[(df['Latitude'] == 0)]

Postcode           Borough                          Neighbourhood  \
25      M3A        North York                              Parkwoods   
85      M7A  Downtown Toronto                           Queen's Park   
86      M7R       Mississauga  Canada Post Gateway Processing Centre   

    Latitude  Longitude                     response  
25       0.0        0.0  Toronto, ON M6K 1X9, Canada  
85       0.0        0.0  Toronto, ON M6K 1X9, Canada  
86       0.0        0.0  Toronto, ON M6K 1X9, Canada

As mentioned above, there are three cases where opencage cannot find the requested postcodes.
That is getting obvious after the opencage response check.  
Therefore, in the code below, the geocodes have been updated manually.

In [10]:
df.loc[df['Postcode'] == 'M3A','Latitude'],df.loc[df['Postcode'] == 'M3A','Longitude']  = 43.7533,-79.3297
df.loc[df['Postcode'] == 'M7A','Latitude'],df.loc[df['Postcode'] == 'M7A','Longitude']  = 43.6623,-79.3895
df.loc[df['Postcode'] == 'M7R','Latitude'],df.loc[df['Postcode'] == 'M7R','Longitude']  = 43.6370,-79.6158
df.drop('response',axis = 1, inplace = True)
df.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.811300   
1      M1C  Scarborough  Highland Creek, Port Union, Rouge Hill  43.787800   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.767800   
3      M1G  Scarborough                                  Woburn  43.765717   
4      M1H  Scarborough                               Cedarbrae  43.768600   

   Longitude  
0 -79.193000  
1 -79.156400  
2 -79.186600  
3 -79.221898  
4 -79.238900

#### Create a map with Toronto neighbourhoods.

In [11]:
%matplotlib notebook
toronto_coord = geocoder.geocode('Toronto, Ontario')
toronto_lat = toronto_coord[0]['geometry']['lat']
toronto_lng = toronto_coord[0]['geometry']['lng']

map_toronto = folium.Map(location = [toronto_lat,toronto_lng],zoom_start = 10)

for lat,lng,borough,neighbourhood in zip(df['Latitude'],df['Longitude'],df['Borough'],df['Neighbourhood']):
    label = '{}: {}'.format(neighbourhood,borough)
    label = folium.Popup(label,parse_html = True)
    folium.CircleMarker([lat,lng],radius = 5,popup = label,color = 'blue',fill = True,
                        fill_color = '#3186cc',fill_opacity = 0.7,parse_html = False).add_to(map_toronto)
map_toronto

# Part III. Exploring the neighbourhoods in Toronto.

## 1. Getting the venues list for each neighbourhood.

In [12]:
# The code was removed by Watson Studio for sharing.

#### Definition of the function "getNearbyVenues". This function is dedicated to getting the venue list (max 100) for each neighbourhood in a 500m radius. 

In [27]:
def getNearbyVenues(neighbourhoods,latitudes,longitudes,radius = 500,limit = 3):    
    venues_list = []
    for neighbourhood,lat,lng in zip(neighbourhoods,latitudes,longitudes):
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID,CLIENT_SECRET,VERSION,lat,lng,radius,limit)
        results = requests.get(url).json()['response']['groups'][0]['items']        
        venues_list.append([(neighbourhood,lat,lng,v['venue']['name'],v['venue']['location']['lat'],v['venue']['location']['lng'],v['venue']['categories'][0]['name'])
                            for v in results])
        
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood','Neighbourhood Latitude','Neighbourhood Longitude','Venue','Venue Latitude','Venue Longitude','Venue Category']
    return(nearby_venues)

#### Getting the venues list for Toronto neighbourhoods.

In [28]:
toronto_venues = getNearbyVenues(neighbourhoods = df['Neighbourhood'],latitudes = df['Latitude'],longitudes = df['Longitude'],limit = 100)

Let's check how many venues were returned for each neighbourhood.

In [29]:
pd.set_option('display.max_rows',1000)
pd.set_option('display.max_columns',1000)
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Agincourt                                                                5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      13   
Alderwood, Long Branch                                                   7   
Bathurst Manor, Downsview North, Wilson Heights                         23   
Bathurst Quay, CN Tower, Harbourfront West, Isl...                     100   
Bayview Village                                                          3   
Bedford Park, Lawrence Manor East                                       25   
Berczy Park                                                             75   
Birch Cliff, Cliffside West                                              4   
Bloordale Gardens, Eringate, Markland Wood, Old...                       1   
Brockton, Exhibition Place, Parkdale Village                            34   
Business Reply Mail Processing Centre 969 Eastern                       15   
CFB Toronto, Downsview East                                              5   
Cabbagetown, St. James Town                                             44   
Caledonia-Fairbanks                                                     10   
Canada Post Gateway Processing Centre                                   12   
Cedarbrae                                                                3   
Central Bay Street                                                      96   
Chinatown, Grange Park, Kensington Market                               76   
Christie                                                                12   
Church and Wellesley                                                    84   
Clairlea, Golden Mile, Oakridge                                         10   
Clarks Corners, Sullivan, Tam O'Shanter                                 13   
Cliffcrest, Cliffside, Scarborough Village West                          9   
Cloverdale, Islington, Martin Grove, Princess G...                      21   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              24   
Davisville North                                                         6   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                       6   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                           3   
Design Exchange, Toronto Dominion Centre                               100   
Don Mills North                                                          4   
Don Mills South, Flemingdon Park                                        42   
Dorset Park, Scarborough Town Centre, Wexford H...                       4   
Dovercourt Village, Dufferin                                            19   
Downsview Central                                                        1   
Downsview Northwest                                                     18   
Downsview West                                                           1   
Downsview, North Park, Upwood Park                                       2   
East Birchmount Park, Ionview, Kennedy Park                             17   
East Toronto                                                             4   
Emery, Humberlea                                                         6   
Fairview, Henry Farm, Oriole                                            12   
First Canadian Place, Underground city                                 100   
Forest Hill North, Forest Hill West                                      3   
Garden District, Ryerson                                               100   
Glencairn                                                               12   
Guildwood, Morningside, West Hill                                       31

#### Short check of the obtained dataframe.

In [30]:
print('Number of unique neighbourhoods: {}'.format(len(toronto_venues['Neighbourhood'].unique())))

Number of unique neighbourhoods: 100


But in the initial dataframe we had 103 neighbourhoods.  
So first, let's find the repetitive neighbourhoods in the initial dataframe:

In [31]:
df[df.duplicated('Neighbourhood',keep = False)]

Postcode           Borough Neighbourhood  Latitude  Longitude
85      M7A  Downtown Toronto  Queen's Park   43.6623   -79.3895
93      M9A      Queen's Park  Queen's Park   43.6662   -79.5282

All right, then one missing row is explained by repeated neighbourhood Queen's Park. How about the other two?  
Let's check API responses for those neighbourhoods, which exist in the initial dataframe, but not in the new one.

In [32]:
for nb in list(set(df['Neighbourhood']) - set(toronto_venues['Neighbourhood'])):
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID,CLIENT_SECRET,VERSION,df.loc[df['Neighbourhood'] == nb,'Latitude'].values[0],df.loc[df['Neighbourhood'] == nb,'Longitude'].values[0],100,500)
    results = requests.get(url).json()['response']['warning']['text']
    print('For neighbourhood = {}: \n {}'.format(nb,results))

For neighbourhood = Upper Rouge: 
 There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area.
For neighbourhood = Highland Creek, Port Union, Rouge Hill: 
 There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area.


So, two disappeared rows are explained by the missing result of the Foursquare API.

## 2. The venue's categories processing.

#### One hot encoding. Let's add venue categories as columns.

In [33]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']],prefix = '',prefix_sep = '')
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 
toronto_onehot = toronto_onehot[[toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])]
toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  \
0                     Malvern, Rouge                  0                  0   
1  Guildwood, Morningside, West Hill                  0                  0   
2  Guildwood, Morningside, West Hill                  0                  0   
3  Guildwood, Morningside, West Hill                  0                  0   
4  Guildwood, Morningside, West Hill                  0                  0   

   Airport  American Restaurant  Antique Shop  Aquarium  Arcade  Art Gallery  \
0        0                    0             0         0       0            0   
1        0                    0             0         0       0            0   
2        0                    0             0         0       0            0   
3        0                    0             0         0       0            0   
4        0                    0             0         0       0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Auto Garage  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  Bakery  \
0            0          0           0                0           0       0   
1            0          0           0                0           0       0   
2            0          0           0                0           0       0   
3            0          0           0                0           0       0   
4            0          0           0                0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0          0         0           0                   0   
1                   0          0         0           0                   0   
2                   0          0         0           0                   0   
3                   0          0         0           0                   0   
4                   0          0         0           0                   0   

   Bike Rental / Bike Share  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                         0       0              0          0         0   
1                         0       0              0          0         0   
2                         0       0              0          0         0   
3                         0       0              0          0         0   
4                         0       0              0          0         0   

   Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0              0                     0               0        0            0   
1              0                     0               0        0            0   
2              0                     0               0        0            0   
3              0                     0               0        0            0   
4              0                     0               0        0            0   

   Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  Bus Line  \
0                0       0         0             0              0         0   
1                0       0         0             0              0         0   
2                0       0         0             0              0     

In [34]:
print('The new dataframe size: {}.'.format(toronto_onehot.shape))

The new dataframe size: (2449, 276).


#### Now, let's group the rows by neighbourhood and by taking the mean of the frequency of the occurrence of each category.

In [35]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                            Adelaide, King, Richmond           0.000000   
1                                           Agincourt           0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...           0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...           0.000000   
4                              Alderwood, Long Branch           0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights           0.000000   
6   Bathurst Quay, CN Tower, Harbourfront West, Is...           0.000000   
7                                     Bayview Village           0.000000   
8                   Bedford Park, Lawrence Manor East           0.000000   
9                                         Berczy Park           0.000000   
10                        Birch Cliff, Cliffside West           0.000000   
11  Bloordale Gardens, Eringate, Markland Wood, Ol...           0.000000   
12       Brockton, Exhibition Place, Parkdale Village           0.000000   
13  Business Reply Mail Processing Centre 969 Eastern           0.000000   
14                        CFB Toronto, Downsview East           0.000000   
15                        Cabbagetown, St. James Town           0.000000   
16                                Caledonia-Fairbanks           0.000000   
17              Canada Post Gateway Processing Centre           0.000000   
18                                          Cedarbrae           0.000000   
19                                 Central Bay Street           0.000000   
20          Chinatown, Grange Park, Kensington Market           0.000000   
21                                           Christie           0.000000   
22                               Church and Wellesley           0.000000   
23                    Clairlea, Golden Mile, Oakridge           0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter           0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West           0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...           0.000000   
27                     Commerce Court, Victoria Hotel           0.000000   
28                                         Davisville           0.000000   
29                                   Davisville North           0.000000   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn           0.000000   
32           Design Exchange, Toronto Dominion Centre           0.000000   
33                                    Don Mills North           0.000000   
34                   Don Mills South, Flemingdon Park           0.000000   
35  Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
36                       Dovercourt Village, Dufferin           0.000000   
37                                  Downsview Central           0.000000   
38                                Downsview Northwest           0.000000   
39                                     Downsview West           0.000000   
40                 Downsview, North Park, Upwood Park           0.000000   
41        East Birchmount Park, Ionview, Kennedy Park           0.000000   
42                                       East Toronto           0.000000   
43                                   Emery, Humberlea           0.000000   
44                       Fairview, Henry Farm, Oriole           0.000000   
45             First Canadian Place, Underground city           0.000000   
46                Forest Hill North, Forest Hill West           0.000000   
47                           Garden District, Ryerson           0.000000   
48                                          Glencairn           0.000000   
49                  Guildwood, Morningside, West Hill           0.000000   
50                     Harbord, University of Toronto           0.000000   
51                                       Harbourfront           0.000000   
52  Harbourf

In [36]:
print('The new dataframe size: {}.'.format(toronto_grouped.shape))

The new dataframe size: (100, 276).


#### Let's print each neighbourhood along with the top 5 most common categories of venues.

In [37]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print('-----' + hood + '-----')
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq':2})
    print(temp.sort_values('freq',ascending = False).reset_index(drop = True).head(num_top_venues))
    print('\n')


-----Adelaide, King, Richmond-----
              venue  freq
0              Café  0.06
1       Coffee Shop  0.06
2             Hotel  0.04
3        Steakhouse  0.04
4  Asian Restaurant  0.03


-----Agincourt-----
                       venue  freq
0             Breakfast Spot   0.2
1            Badminton Court   0.2
2        Shanghai Restaurant   0.2
3  Latin American Restaurant   0.2
4               Skating Rink   0.2


-----Agincourt North, L'Amoreaux East, Milliken, Steeles East-----
                        venue  freq
0                    Pharmacy   0.5
1            Sushi Restaurant   0.5
2           Accessories Store   0.0
3  Modern European Restaurant   0.0
4                  Nail Salon   0.0


-----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown-----
            venue  freq
0   Grocery Store  0.15
1     Pizza Place  0.08
2    Liquor Store  0.08
3  Sandwich Place  0.08
4      Beer Store  0.08


-----Alderwood, Long Bra

#### Let's write a function to sort the venues in descending order.

In [38]:
def return_most_common_venues(row,num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now let's create the new dataframe and display the top 10 venues for each neighbourhood.

In [39]:
num_top_venues = 10
indicators = ['st','nd','rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind + 1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind + 1))

neighbourhoods_venues_sorted = pd.DataFrame(columns = columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind,1:] = return_most_common_venues(toronto_grouped.iloc[ind,:], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East, Milliken, St...              Pharmacy   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Café            Steakhouse                 Hotel   
1  Latin American Restaurant       Badminton Court   Shanghai Restaurant   
2           Sushi Restaurant           Yoga Studio   Dumpling Restaurant   
3                Pizza Place           Auto Garage        Hardware Store   
4             Sandwich Place           Coffee Shop     Convenience Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                Bakery                   Gym   
1          Skating Rink           Event Space  Ethiopian Restaurant   
2    Dim Sum Restaurant                 Diner        Discount Store   
3           Coffee Shop  Caribbean Restaurant        Sandwich Place   
4                   Pub              Pharmacy                   Gym   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Asian Restaurant        Breakfast Spot           Burger Joint  
1     Electronics Store    Falafel Restaurant     Dim Sum Restaurant  
2               Dog Run      Doner Restaurant             Donut Shop  
3  Fast Food Restaurant          Liquor Store             Beer Store  
4      Greek Restaurant           Curling Ice    Dumpling Restaurant

## 3. The clustering of the neighbourhoods, based on their venues categories representation.

#### To cluster neighbourhoods, we will be using the K-Means clustering.

In [40]:
kclusters = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood',1)
kmeans = KMeans(n_clusters = kclusters,random_state = 0).fit(toronto_grouped_clustering)
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 2, 0, 0], dtype=int32)

#### Now, let's add cluster labels, as well as the most common venues, into the initial dataframe.

In [41]:
neighbourhoods_venues_sorted.insert(0,'Cluster Labels',kmeans.labels_)
toronto_merged = df
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'),on = 'Neighbourhood')
toronto_merged.dropna(inplace = True)
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)
toronto_merged.head()

Postcode      Borough                      Neighbourhood   Latitude  \
0      M1B  Scarborough                     Malvern, Rouge  43.811300   
2      M1E  Scarborough  Guildwood, Morningside, West Hill  43.767800   
3      M1G  Scarborough                             Woburn  43.765717   
4      M1H  Scarborough                          Cedarbrae  43.768600   
5      M1J  Scarborough                Scarborough Village  43.746400   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.193000               3          Home Service           Yoga Studio   
2 -79.186600               0           Pizza Place  Fast Food Restaurant   
3 -79.221898               0            Taco Place     Indian Restaurant   
4 -79.238900               0                Lounge                 Trail   
5 -79.232300               0                   Spa         Grocery Store   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Dumpling Restaurant    Dim Sum Restaurant                 Diner   
2               Grocery Store              Pharmacy           Coffee Shop   
3         American Restaurant                  Park           Coffee Shop   
4  Construction & Landscaping   Dumpling Restaurant                 Diner   
5                 Yoga Studio   Dumpling Restaurant    Dim Sum Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store               Dog Run      Doner Restaurant   
2        Discount Store          Intersection     Food & Drink Shop   
3           Yoga Studio                 Diner        Discount Store   
4        Discount Store               Dog Run      Doner Restaurant   
5                 Diner        Discount Store               Dog Run   

  9th Most Common Venue       10th Most Common Venue  
0            Donut Shop  Eastern European Restaurant  
2            Beer Store                  Supermarket  
3               Dog Run             Doner Restaurant  
4            Donut Shop  Eastern European Restaurant  
5      Doner Restaurant                   Donut Shop

#### And now, it's time to visualize the resulting clusters!

In [42]:
map_clusters = folium.Map(location = [toronto_lat,toronto_lng],zoom_start = 10)

x = np.arange(kclusters)
ys = [i + x + (i * x) ** 2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0,1,len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat,lon,poi,cluster in zip(toronto_merged['Latitude'],toronto_merged['Longitude'],toronto_merged['Neighbourhood'],toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster),parse_html = True)
    folium.CircleMarker([lat,lon],radius = 5,popup = label,color = rainbow[cluster-1],fill = True,fill_color = rainbow[cluster-1],fill_opacity = 0.7).add_to(map_clusters)
map_clusters

### Let's have a look at the clusters one by one.

#### Cluster 1.

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels        1st Most Common Venue  \
2         Scarborough               0                  Pizza Place   
3         Scarborough               0                   Taco Place   
4         Scarborough               0                       Lounge   
5         Scarborough               0                          Spa   
6         Scarborough               0                  Coffee Shop   
7         Scarborough               0                       Bakery   
8         Scarborough               0               Ice Cream Shop   
9         Scarborough               0                 Skating Rink   
10        Scarborough               0   Construction & Landscaping   
12        Scarborough               0               Breakfast Spot   
13        Scarborough               0                  Pizza Place   
14        Scarborough               0                     Pharmacy   
15        Scarborough               0           Chinese Restaurant   
20         North York               0            Martial Arts Dojo   
22         North York               0            Korean Restaurant   
27         North York               0                 Gourmet Shop   
28         North York               0                  Coffee Shop   
29         North York               0    Middle Eastern Restaurant   
30         North York               0                      Airport   
31         North York               0                         Pool   
33         North York               0                Grocery Store   
34         North York               0                 Intersection   
35          East York               0                  Pizza Place   
36          East York               0                   Beer Store   
37       East Toronto               0            Health Food Store   
38          East York               0            Electronics Store   
39          East York               0            Indian Restaurant   
41       East Toronto               0             Greek Restaurant   
42       East Toronto               0                  Coffee Shop   
43       East Toronto               0                        Diner   
47    Central Toronto               0                  Coffee Shop   
49    Central Toronto               0           Light Rail Station   
51   Downtown Toronto               0                  Coffee Shop   
52   Downtown Toronto               0                  Coffee Shop   
53   Downtown Toronto               0                  Coffee Shop   
54   Downtown Toronto               0                  Coffee Shop   
55   Downtown Toronto               0                  Coffee Shop   
56   Downtown Toronto               0                  Coffee Shop   
57   Downtown Toronto               0                  Coffee Shop   
58   Downtown Toronto               0                  Coffee Shop   
59   Downtown Toronto               0                  Coffee Shop   
60   Downtown Toronto               0                  Coffee Shop   
61   Downtown Toronto               0                  Coffee Shop   
62         North York               0         Fast Food Restaurant   
63    Central Toronto               0      Health & Beauty Service   
64    Central Toronto               0                     Bus Line   
65    Central Toronto               0                         Café   
66   Downtown Toronto               0                         Café   
67   Downtown Toronto               0                          Bar   
68   Downtown Toronto               0                        Hotel   
69   Downtown Toronto               0                  Coffee Shop   
70   Downtown Toronto               0                  Coffee Shop   
71         North York               0               Clothing Store   
72         North York               0                  Pizza Place   
75   Downtown Toronto               0                         Café   
76       West Toronto               0                       Bakery   
77       West Toronto   

#### Cluster 2.

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Scarborough               1           Auto Garage           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11      Department Store    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11        Discount Store               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
11            Donut Shop    Dumpling Restaurant

#### Cluster 3.

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17        North York               2                  Park   
18        North York               2                  Park   
19        North York               2           Gas Station   
21        North York               2            Playground   
23        North York               2     Convenience Store   
24        North York               2            Baby Store   
25        North York               2     Food & Drink Shop   
26        North York               2      Basketball Court   
40         East York               2     Convenience Store   
44   Central Toronto               2    Photography Studio   
45   Central Toronto               2      Department Store   
46   Central Toronto               2                Garden   
48   Central Toronto               2                  Park   
50  Downtown Toronto               2           Candy Store   
73              York               2         Deli / Bodega   
74              York               2                  Park   
81              York               2               Brewery   
91         Etobicoke               2        Baseball Field   

                       2nd Most Common Venue 3rd Most Common Venue  \
17  Residential Building (Apartment / Condo)           Yoga Studio   
18                                  Pharmacy           Pizza Place   
19                                     Trail                  Park   
21                                       Gym           Yoga Studio   
23                                      Park                  Bank   
24                                      Park           Coffee Shop   
25                                      Park           Yoga Studio   
26                Construction & Landscaping                  Pool   
40                                      Park           Coffee Shop   
44                                      Park           Yoga Studio   
45                            Breakfast Spot               Dog Run   
46                                Playground              Gym Pool   
48                              Tennis Court                   Gym   
50                                Playground         Grocery Store   
73                                     Field            Playground   
74                                    Bakery   Sporting Goods Shop   
81                    Furniture / Home Store                  Park   
91                                      Park           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
17      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
18             Wine Shop           Coffee Shop       Thai Restaurant   
19           Yoga Studio      Doner Restaurant          Dessert Shop   
21   Dumpling Restaurant    Dim Sum Restaurant                 Diner   
23           Yoga Studio   Dumpling Restaurant                 Diner   
24           Yoga Studio            Donut Shop    Dim Sum Restaurant   
25   Dumpling Restaurant    Dim Sum Restaurant                 Diner   
26                  Park           Yoga Studio            Donut Shop   
40      Greek Restaurant           Yoga Studio   Dumpling Restaurant   
44            Donut Shop          Dessert Shop    Dim Sum Restaurant   
45                   Gym                  Park     Food & Drink Shop   
46                  Park      Doner Restaurant      Department Store   
48         Grocery Store            Playground       Thai Restaurant   
50                  Park      Doner Restaurant      Department Store   
73         Grocery Store          Hockey Arena                  Park   
74                Market    Mexican Restaurant            Beer Store   
81           Yoga Studio   Dumpling Restaurant                 Diner   
91   Dumpling Restaurant                 Diner        Discount Store   

   7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
17                 Diner             Discount Store      

#### Cluster 4.

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Scarborough               3          Home Service           Yoga Studio   
96   North York               3          Home Service           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Dumpling Restaurant    Dim Sum Restaurant                 Diner   
96   Dumpling Restaurant    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Discount Store               Dog Run      Doner Restaurant   
96        Discount Store               Dog Run      Doner Restaurant   

   9th Most Common Venue       10th Most Common Venue  
0             Donut Shop  Eastern European Restaurant  
96            Donut Shop  Eastern European Restaurant

#### Cluster 5.

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  North York               4        Baseball Field           Yoga Studio   
95   Etobicoke               4        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32   Dumpling Restaurant                 Diner        Discount Store   
95   Dumpling Restaurant                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32               Dog Run      Doner Restaurant            Donut Shop   
95               Dog Run      Doner Restaurant            Donut Shop   

          9th Most Common Venue 10th Most Common Venue  
32  Eastern European Restaurant           Dessert Shop  
95  Eastern European Restaurant           Dessert Shop

# Many thanks for your time!  
# The end.